src/librustc_mir/transform/generator.rs

```diff
@@ -130,9 +130,9 @@ impl<'tcx> MutVisitor<'tcx> for PinArgVisitor<'tcx> {
                     place: &mut Place<'tcx>,
                     context: PlaceContext,
                     location: Location) {
-        if *place == Place::Base(PlaceBase::Local(self_arg())) {
-            *place = Place::Projection(Box::new(Projection {
-                base: place.clone(),
+        if *place == Place::from(self_arg()) {
```

Comment:
this (and all others in this file), should be `if place.base == PlaceBase::Local(self_arg())`, but the change should not be applied to the outermost projection, but to the innermost one. You could create a new free function in this file that replaces the innermost `None` projetion with `Some(Box::new(Projection { base: None, elem: argument_to_new_free_func }))`